### PR TITLE
Move helper methods that should really exist on the model

### PR DESCRIPTION
**Pull request by joshlory**
This is a prerequisite for merging https://github.com/code-dot-org/code-dot-org/pull/1532.  Pulled and squashed from branch:api-changes.

### CODE REVIEW

**Comment by wjordan** [comment]
Personally, I think I'd prefer to keep i18n functionality within the view layer (e.g. within the view helpers they were in previously), so that the i18n view-rendering logic stays separate from the data-modeling and validation layer. (maybe worth following up with @laurelfan for a second opinion)

In any case, since this is a prerequisite for other existing changes to come, the code changes look fine and I don't think the MVC-architecture discussion needs to hold up work at this point.

**Comment by joshlory** [comment]
@laurelfan?  We moved localization of `.summarize` to the model because there will be multiple consumers of this API.  While each consumer could manually localize the strings coming from the model, it feels cleaner to localize it once at the model layer.

**Comment by joshlory** [comment]
(One of these consumers will be a JSON API, so localizing in the view isn't possible yet without a front-end i18n module.)

File: dashboard/app/models/video.rb [hunk 1]
```diff
@@ -21,4 +21,47 @@ def self.setup
     end
     check_i18n_names
   end
+
+  def self.youtube_base_url
+    'https://www.youtube.com'
+  end
+
+  def self.youtube_url(code, args={})
```
Comment (laurelfan) [inline dashboard/app/models/video.rb]: where does code come from? Is it a property of the Video model and if so, should this be an instance method that gets code from self?

**Comment by laurelfan** [comment]
I think this is where two reasonable 'rules' conflict, where those rules are:
1. things relating to a single model object belong in the model
2. things relating to rendering belong in views and view helpers

It seems that with the jsonification of things we're moving 'view' stuff closer to the models (since json is a view...). (side note, the serializer stuff that @wjordan is using in the ops side of things is another example of a different way of handling things that are viewy and modely).

Personally, my opinion is that having methods that don't use self and have one argument are an indication that they should be instance methods on that argument's class (assuming they don't then need to have `request` passed in or anything like that). So these changes make the code look better to me.

It also seems to be a matter of opinion whether i18n is a "rendering" thing. In the past we've been acting as if it is (mostly) and doing that in the view, but I think I18n.t being available in the model indicates that this is not necessarily the way we have to do it.

File: dashboard/app/models/video.rb (hunk 1 above)
Comment (joshlory) [inline dashboard/app/models/video.rb]: Good catch.  Now that it's in the model I can refactor it to not take `code` as a param.

**Comment by SteveEisner** [comment]
FWIW I also balked at the idea of i18n (and a few other things) moving into the models at first!  But I think you can justify i18n in model as a "half step" towards eventually sending un-i18n'd tokens to the client side.  And I like laurel's explanation of conflicting principles...

re: jsonification, my mental model of JSON is really just "plain old javascript objects that happen to be able to be sent across the wire".  Where it's spreading on the server you might think of it as a "view" but you might also think of it as just a non-ORM'd model, and some people use the term "viewmodel".  There are interesting properties of viewmodels that apply nicely to the server as much as the client, for instance they are read-only 'normalized' views that hide complexity, promise cheap member access (vs possible overhead of methods), are ideal as API returns, and so on.

Of course they can also introduce complexity because you now have to think about whether you're working with a script model or a script viewmodel, and write different code.  And if they grow big enough then they start to need some of the same overhead that regular models do (lazy/partial loading and so on.)

I think it works out well in this code but I will say that Ruby makes it harder to reference fields --myobject[:property] instead of myobject.property has really frustrated me when writing the code :(

**Comment by wjordan** [comment]
Thanks for the extra discussion all, I'm glad we've at least thought this through somewhat (and as I mentioned earlier, I think the code is good to go and don't think any of this discussion should block any merge when it's ready).

I still stick to my recommendation that we eventually try to keep `I18n.t`-calling code out of the model classes. I don't think it's a matter of opinion that I18n is a 'rendering' thing. Regardless of where the method is placed, I18n-localized strings are being rendered through the request-processing controller layer, through the global state of the user's current locale (passed around in the global `I18n.locale` variable).

The result of a `I18n.t` call depends on the current value of `I18n.locale`, as manipulated by the controller layer depending on the content of request headers, session objects, Varnish-filtered cookies, etc. This means that we have to make sure that we _never_ depend on strings being returned from `I18n.t`-using methods being consistent from one call to the next, regardless of whether the model object itself changes state. As soon as a string fed through an `I18n` call gets accidentally fed into some state cached in a class-variable for example, we have a nasty i18n bug. We tend to do quite a bit of caching logic in our model objects, so I worry that these kinds of bugs will appear quite quickly.

Also, as Laurel touched upon, I would much rather we think of JSON APIs as providing a more lightweight (but still structured) view on the underlying model object, rather than pushing this leftover logic back down onto the model.

**Comment by joshlory** [comment]
@wjordan you make a great point about the potential bugs from relaying on `I18n.locale` across requests.  Should we roll back this change until we can better address this issue?  If it's going to break, it will break quickly on a DTP since the title of the scripts and stages are used many places across the live site.

**Comment by wjordan** [comment]
I don't think this merge itself will present any problem, I see no reason to revert - I just wanted to raise the architectural issue as something we can make sure to think carefully about and address moving forward.

**Pull request merged**